### DISCUSSION TITLE
MQTT error: connect ECONNREFUSED

### DISCUSSION THREAD

**Discussion by bremenpl** [Koenkk/zigbee2mqtt#19021] [General]
Hello,
I have installed the zigbee2mqtt via this tutorial: https://flemmingss.com/how-to-set-up-zigbee2mqtt-on-a-raspberry-pi-and-integrate-it-with-home-assistant/

The only change I did is using `https://deb.nodesource.com/setup_20.x` instead of `https://deb.nodesource.com/setup_14.x`

I intend to have the mqtt server on a raspberry pi and connect to it from a remote home assistant instance. The problem I am facing is that the service is failing. This is a part from the journal:

```
Sep 19 16:15:27 rpi3 systemd[1]: Started zigbee2mqtt.
Sep 19 16:15:34 rpi3 npm[545]: > zigbee2mqtt@1.33.0 start
Sep 19 16:15:34 rpi3 npm[545]: > node index.js
Sep 19 16:15:44 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:44: Logging to console and directory: '/opt/zigbee2mqtt/data/log/2023-09-19.16-15-44' filen>
Sep 19 16:15:44 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:44: Starting Zigbee2MQTT version 1.33.0 (commit #f224ef7b)
Sep 19 16:15:44 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:44: Starting zigbee-herdsman (0.18.5)
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:49: zigbee-herdsman started (resumed)
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:49: Coordinator firmware version: '{"meta":{"maintrel":1,"majorrel":2,"minorrel":7,"product>
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:49: Currently 0 devices are joined:
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:warn  2023-09-19 16:15:49: `permit_join` set to  `true` in configuration.yaml.
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:warn  2023-09-19 16:15:49: Allowing new devices to join.
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:warn  2023-09-19 16:15:49: Set `permit_join` to `false` once you joined all devices.
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:49: Zigbee: allowing new devices to join.
Sep 19 16:15:49 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:49: Connecting to MQTT server at mqtt://192.168.7.235
Sep 19 16:15:50 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:50: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:50 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:50: MQTT failed to connect, exiting...
Sep 19 16:15:50 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:15:50: Stopping zigbee-herdsman...
Sep 19 16:15:51 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:51: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:52 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:52: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:53 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:53: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:54 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:54: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:55 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:55: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:15:56 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:15:56: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:16:12 rpi3 npm[652]: Zigbee2MQTT:error 2023-09-19 16:16:12: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 19 16:16:12 rpi3 npm[652]: Zigbee2MQTT:info  2023-09-19 16:16:12: Stopped zigbee-herdsman
Sep 19 16:16:13 rpi3 systemd[1]: zigbee2mqtt.service: Main process exited, code=exited, status=1/FAILURE
Sep 19 16:16:13 rpi3 systemd[1]: zigbee2mqtt.service: Failed with result 'exit-code'.
Sep 19 16:16:13 rpi3 systemd[1]: zigbee2mqtt.service: Consumed 29.126s CPU time.
Sep 19 16:16:13 rpi3 systemd[1]: zigbee2mqtt.service: Scheduled restart job, restart counter is at 1.
Sep 19 16:16:13 rpi3 systemd[1]: Stopped zigbee2mqtt.
Sep 19 16:16:13 rpi3 systemd[1]: zigbee2mqtt.service: Consumed 29.126s CPU time.
Sep 19 16:16:13 rpi3 systemd[1]: Started zigbee2mqtt.
Sep 19 16:16:15 rpi3 npm[761]: > zigbee2mqtt@1.33.0 start
Sep 19 16:16:15 rpi3 npm[761]: > node index.js
```

`192.168.7.235` is the IP address of the RPI the service is running on. This is my config file:

```
pi@rpi3:~ $ cat /opt/zigbee2mqtt/data/configuration.yaml
homeassistant: true
permit_join: true
mqtt:
  base_topic: zigbee2mqtt
  server: 'mqtt://192.168.7.235'
serial:
  port: /dev/ttySonof
advanced:
  network_key:
    - 88
    - 138
    - 225
    - 61
    - 6
    - 122
    - 99
    - 95
    - 211
    - 124
    - 38
    - 99
    - 247
    - 204
    - 44
    - 92
  homeassistant_legacy_entity_attributes: false
  legacy_api: false
  legacy_availability_payload: false
device_options:
  legacy: false

frontend: true
```

I cannot figure out why it is failing. I would appreciate all help!

**Comment by francisp2** [comment]
Why don't you add a user and password to both your mosquitto and Zigbee2MQTT configuration ? Mosquitto 2.x refuses anonymous connections, unless explicitly allowed in its configuration.

**Reply by bremenpl** [reply]
Hi @francisp2 , thank you for the answer. I tried adding user and password to the `/opt/zigbee2mqtt/data/configuration.yaml`, but that did not change the case. Where else ("mosquito") should I add the credentials?

**Reply by francisp2** [reply]
https://mosquitto.org/documentation/authentication-methods/

**Reply by bremenpl** [reply]
I have added the following to the `/opt/zigbee2mqtt/data/configuration.yaml`:

```
listener: 1883
allow_anonymous: true
```
But it seems that it is not the case, as the error is still the same. 

```
Sep 20 10:49:36 rpi3 systemd[1]: Started zigbee2mqtt.
Sep 20 10:49:43 rpi3 npm[548]: > zigbee2mqtt@1.33.0 start
Sep 20 10:49:43 rpi3 npm[548]: > node index.js
Sep 20 10:49:53 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:53: Logging to console and directory: '/opt/zigbee2mqtt/data/log/2023-09-20.10-49-53' filen>
Sep 20 10:49:54 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:54: Starting Zigbee2MQTT version 1.33.0 (commit #f224ef7b)
Sep 20 10:49:54 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:54: Starting zigbee-herdsman (0.18.5)
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:58: zigbee-herdsman started (resumed)
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:58: Coordinator firmware version: '{"meta":{"maintrel":1,"majorrel":2,"minorrel":7,"product>
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:58: Currently 0 devices are joined:
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:warn  2023-09-20 10:49:58: `permit_join` set to  `true` in configuration.yaml.
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:warn  2023-09-20 10:49:58: Allowing new devices to join.
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:warn  2023-09-20 10:49:58: Set `permit_join` to `false` once you joined all devices.
Sep 20 10:49:58 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:58: Zigbee: allowing new devices to join.
Sep 20 10:49:59 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:59: Connecting to MQTT server at mqtt://192.168.7.235
Sep 20 10:49:59 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:49:59: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:49:59 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:49:59: MQTT failed to connect, exiting...
Sep 20 10:49:59 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:49:59: Stopping zigbee-herdsman...
Sep 20 10:50:00 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:00: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:01 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:01: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:02 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:02: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:03 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:03: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:04 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:04: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:05 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:05: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:20 rpi3 npm[693]: Zigbee2MQTT:error 2023-09-20 10:50:20: MQTT error: connect ECONNREFUSED 192.168.7.235:1883
Sep 20 10:50:21 rpi3 npm[693]: Zigbee2MQTT:info  2023-09-20 10:50:21: Stopped zigbee-herdsman
Sep 20 10:50:21 rpi3 systemd[1]: zigbee2mqtt.service: Main process exited, code=exited, status=1/FAILURE
Sep 20 10:50:21 rpi3 systemd[1]: zigbee2mqtt.service: Failed with result 'exit-code'.
Sep 20 10:50:21 rpi3 systemd[1]: zigbee2mqtt.service: Consumed 29.219s CPU time.
Sep 20 10:50:22 rpi3 systemd[1]: zigbee2mqtt.service: Scheduled restart job, restart counter is at 1.
Sep 20 10:50:22 rpi3 systemd[1]: Stopped zigbee2mqtt.
Sep 20 10:50:22 rpi3 systemd[1]: zigbee2mqtt.service: Consumed 29.219s CPU time.
```

Is there any way to make the log more verbose about its issues?

**Comment by francisp2** [comment]
You have to add 
```
allow_anonymous true
```
to your mosquitto.conf file and restart mosquitto. 

> I have added the following to the /opt/zigbee2mqtt/data/configuration.yaml:
>
>listener: 1883
>allow_anonymous: true

does nothing and needs to be removed again

**Reply by bremenpl** [reply]
_does nothing and needs to be removed again_

Ok, removed.

_to your mosquitto.conf file and restart mosquitto._

Where can this file be found? I cannot find it in `/opt/zigbee2mqtt/`.

**Reply by lux73** [reply]
the `mosquitto.conf` File can be found in your Mosquitto Installation Folder

**Comment by francisp2** [comment]
Normally (Debian) in /etc/mosquitto

**Reply by bremenpl** [reply]
That's the thing- I haven't installed Mosquito on the RPI I am running the zigbee2mqtt- at least I weren't instructed to in here: https://flemmingss.com/how-to-set-up-zigbee2mqtt-on-a-raspberry-pi-and-integrate-it-with-home-assistant/

I only have a Mosquito Broker on a remote Home Assistant machine. I feel like I am missing some crucial info in here... Is the server supposed to be installed on localhost?

**Comment by francisp2** [comment]
If you are running the HA Mosquitto broker add-on, you need to create a user in HA, and add that user and password to Zigbee2MQTT's configuration.yaml

**Reply by bremenpl** [reply]
So I understand that:

In this file `/opt/zigbee2mqtt/data/configuration.yaml` I set the `server` to the HA device server, not localhost? And this:

```
  # MQTT server authentication, uncomment if required:
  user: mqtt
  password: password123
```

Is actually the user and password I create for the user in HA Addon?

**Comment by francisp2** [comment]
Yes

**Reply by bremenpl** [reply]
Ok, thank you for the info! In this case I think this wont work for me... In my HA I am already subscribed to one mqtt sourxe and AFAIK this is a limitation of the mqtt broker addon...

**Reply by lux73** [reply]
You only need one (1) mqtt Broker (mosquitto)

All other services (mqtt Clients) are able to connect to this broker

This worked as designed

**Reply by bremenpl** [reply]
Ok, I am clearly missing a point. So in my HA settings I have the MQQT plugin already installed and configured to subscribe to data from supla.org:

![image](https://github.com/Koenkk/zigbee2mqtt/assets/1183500/5e2f735c-9692-4343-8efa-9e9009a99baa)

Doesn't that mean I can no longer use the zigbe2mqtt with this HA instance? I do have a password and a username here, but its related with the supla.org server. Can it be reused for the zigbee2mqtt purposes...?

I would appreciate your further feedback and thank you for your patience.

**Reply by lux73** [reply]
mqtt85.supla.org?

please do a ping to the Adress from your z2m Host and show output

your mqtt broker is running inside your Network and should be available via CIDR IP Adress e.g. 10.x.x.x or 172.16.x.x or at least 192.168.x.x

You should not use any DNS Name

**Reply by bremenpl** [reply]
I am connecting to that server to poll data from it. Like this: https://en-forum.supla.org/viewtopic.php?t=12253

**Reply by lux73** [reply]
I could only strong recommend to setup your own mosquitto mqtt broket

At the Moment you are using an indifferent service accessing via Inet

If your Inet Connection is offline you are not able to pull any Data anymore - even in my opinion its unsecure

**Reply by bremenpl** [reply]
In general, AFAIK, this is the only way to link suppla related data to HA... In this config, I understand it is not possible to directly use the zigbe2mqqt? The only thing here that comes up to my mind is to aggregate different data sources via external broker and subscribe to it in HA?

**Reply by lux73** [reply]
Your "suppla" refers actually as your mqtt broker

In my opinion this is not what you want - why you want use this external service?

If your HA instance not brining any mqtt broker - install another docker container that includes mosquitto 😉

Keep mind that you want the "KISS" priciple - Keep It Stupid Simple

Any other services thats not "under your Hood" complicate things

**Reply by bremenpl** [reply]
I simply followed a tutorial for the supla connection. I guess this then requires more config/ understanding of the topic to resolve. Thanks!

**Reply by francisp2** [reply]
Not recommended, but you could change Zigbee2MQTT's configuration.yaml like this: 
```
mqtt:
  base_topic: zigbee2mqtt
  server: mqtts://mqtt85.supla.org:8883
  user: 07df...
  password: ***
```
Better way is to create a mqtt between a local broker and your supla broker.

**Reply by bremenpl** [reply]
Will the zigbee2mqtt entries be available then via supla...?

**Reply by bremenpl** [reply]
Wow, this worked:

```
-- Boot fcd97db65e2f4e63a8387a3df00b1d55 --
Sep 22 11:52:25 rpi3 systemd[1]: Started zigbee2mqtt.
Sep 22 11:52:30 rpi3 npm[14797]: > zigbee2mqtt@1.33.0 start
Sep 22 11:52:30 rpi3 npm[14797]: > node index.js
Sep 22 11:52:40 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:40: Logging to console and directory: '/opt/zigbee2mqtt/data/log/2023-09-22.11-52-40' filena>
Sep 22 11:52:40 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:40: Starting Zigbee2MQTT version 1.33.0 (commit #f224ef7b)
Sep 22 11:52:40 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:40: Starting zigbee-herdsman (0.18.5)
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:45: zigbee-herdsman started (resumed)
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:45: Coordinator firmware version: '{"meta":{"maintrel":1,"majorrel":2,"minorrel":7,"product">
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:45: Currently 0 devices are joined:
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:warn  2023-09-22 11:52:45: `permit_join` set to  `true` in configuration.yaml.
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:warn  2023-09-22 11:52:45: Allowing new devices to join.
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:warn  2023-09-22 11:52:45: Set `permit_join` to `false` once you joined all devices.
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:45: Zigbee: allowing new devices to join.
Sep 22 11:52:45 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:45: Connecting to MQTT server at mqtts://mqtt85.supla.org:8883
Sep 22 11:52:46 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:46: Connected to MQTT server
Sep 22 11:52:46 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:46: MQTT publish: topic 'zigbee2mqtt/bridge/state', payload '{"state":"online"}'
Sep 22 11:52:46 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:46: Started frontend on port 192.168.7.235:8080
Sep 22 11:52:46 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:46: MQTT publish: topic 'zigbee2mqtt/bridge/state', payload '{"state":"online"}'
Sep 22 11:52:46 rpi3 npm[14813]: Zigbee2MQTT:info  2023-09-22 11:52:46: Zigbee2MQTT started!

```

I really do need to understand this mqtt architecture better. Thank you a lot.

**Reply by bremenpl** [reply]
While I was able to add a device via frontend:

![image](https://github.com/Koenkk/zigbee2mqtt/assets/1183500/2c8cb9f8-7e72-40cb-9278-7fd30b10f82c)

I am still not sure how to make this available to the HA.

**Reply by francisp2** [reply]
Add
homeassistant: true

to the configuration.yaml, and restart Z2M

**Reply by bremenpl** [reply]
The integration is enabled:

![image](https://github.com/Koenkk/zigbee2mqtt/assets/1183500/920f0861-2de8-4a82-9f4c-3b3de6d6f26d)

But in HA I still see only suppla related devices:

![image](https://github.com/Koenkk/zigbee2mqtt/assets/1183500/cd1bd252-8d32-4864-ad37-1f358864c323)

**Reply by lux73** [reply]
I'm using OpenHAB - so i couldnt help

But generally this is not a z2m Problem anymore - i suggest ask HA Community Forum

**Reply by bremenpl** [reply]
Noted, thank you for your help!

**Reply by bremenpl** [reply]
@lux73 , do you think this would work?
1. Disconnect suppla broker from HA,
2. Install Mosquito on rpi,
3. Link zigbee2mqtt to Mosquito,
4. Create a mosquito bridge to connect suppla broker,
5. Connect HA to Mosquito.

Hopefully this way I have both supla and mosquito brokers data...?

**Reply by lux73** [reply]
Could work, give it a try 😎
